src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -413,16 +326,139 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
+                Checksum solutionChecksum, Func<CancellationToken, Task<Solution>> getSolutionAsync, CancellationToken cancellationToken)
+            {
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    if (_solutionChecksumToLazySolution.TryGetValue(solutionChecksum, out var refCountedLazySolution))
+                    {
+                        var lazySolutionInstance = refCountedLazySolution.TryAddReference();
+                        if (lazySolutionInstance is not null)
+                            return lazySolutionInstance;
+
+                        // Remove the value since it's clearly no longer usable. The cleanupAsync method would have
+                        // removed this value, but has not completed its execution yet.
+                        _solutionChecksumToLazySolution.Remove(solutionChecksum);
+                    }
+
+                    // We're the first call that is asking about this checksum.  Create a lazy to compute it with a
+                    // refcount of 1 (for 'us').
+                    refCountedLazySolution = null;
+                    var lazySolution = new LazySolution(
+                        getSolutionAsync,
+                        cleanupAsync: async () =>
+                        {
+                            // We use CancellationToken.None here as we have to ensure the lazy solution is removed from the
+                            // checksum map, or else we will have a memory leak.  This should hopefully not ever be an issue as we
+                            // only ever hold this gate for very short periods of time in order to set do basic operations on our
+                            // state.
+                            using var _ = await _gate.DisposableWaitAsync(CancellationToken.None).ConfigureAwait(false);
+
+                            // Only remove a value from the map if it still exists and holds the same expected instance
+                            if (_solutionChecksumToLazySolution.TryGetValue(solutionChecksum, out var remainingRefCountedLazySolution)
+                                && remainingRefCountedLazySolution == refCountedLazySolution)
+                            {
+                                _solutionChecksumToLazySolution.Remove(solutionChecksum);
+                            }
+                        });
+                    refCountedLazySolution = new ReferenceCountedDisposable<LazySolution>(lazySolution);
+                    _solutionChecksumToLazySolution.Add(solutionChecksum, refCountedLazySolution);
+                    return refCountedLazySolution;
+                }
+            }
+
+            public async Task SetLastRequestedSolutionAsync(Checksum solutionChecksum, ReferenceCountedDisposable<LazySolution> solution, CancellationToken cancellationToken)
+            {
```

Comment:
same as before, just simpler as there's only one 'last requested' solution to update.